cpp/src/arrow/ipc/message.h

```diff
@@ -163,6 +165,135 @@ class ARROW_EXPORT Message {
+
+  virtual ~MessageReceiverAssign() = default;
+
+  Status Received(std::unique_ptr<Message> message) override {
+    *message_ = std::move(message);
+    return Status::OK();
+  }
+
+ private:
+  std::unique_ptr<Message>* message_;
+
+  ARROW_DISALLOW_COPY_AND_ASSIGN(MessageReceiverAssign);
+};
+
+/// \class MessageEmitter
+/// \brief Push style message reader that receives data from user.
+class ARROW_EXPORT MessageEmitter {
+ public:
+  /// \brief State for reading a message
+  enum State {
+    /// The initial state. It requires one of the followings as the next data:
+    ///
+    ///   * int32_t continuation token
+    ///   * int32_t end-of-stream mark (== 0)
+    ///   * int32_t metadata length (backward compatibility for
+    ///     reading old IPC messages produced prior to version 0.15.0
+    INITIAL,
+
+    /// It requires int32_t metadata length.
+    METADATA_LENGTH,
+
+    /// It requires metadata.
+    METADATA,
+
+    /// It requires message body.
+    BODY,
+
+    /// The end-of-stream state. No more data is processed.
+    EOS,
+  };
+
+  /// \brief Construct a message emitter.
+  ///
+  /// \param[in] receiver a MessageReceiver that receives read messages
+  /// \param[in] pool an optional MemoryPool to copy metadata on the
+  /// CPU, if required
+  explicit MessageEmitter(MessageReceiver* receiver,
+                          MemoryPool* pool = default_memory_pool());
+
+  /// \brief Construct a message emitter with the specified state.
+  ///
+  /// This is a construct for advanced users that know how to read
+  /// Message.
+  ///
+  /// \param[in] receiver a MessageReceiver that receives read messages
+  /// \param[in] initial_state an initial state of the emitter
+  /// \param[in] initial_next_required_size an initial size to run
+  /// the next action
+  /// \param[in] pool an optional MemoryPool to copy metadata on the
+  /// CPU, if required
+  MessageEmitter(MessageReceiver* receiver, State initial_state,
+                 int64_t initial_next_required_size,
+                 MemoryPool* pool = default_memory_pool());
+
+  virtual ~MessageEmitter();
+
+  /// \brief Feed data to the emitter as a raw data.
+  ///
+  /// If the emitter can read one or more messages by the data, the
+  /// emitter emits read batches by calling receiver->Receive()
+  /// multiple times.
+  ///
+  /// \param[in] data a raw data to be processed. This data isn't
+  /// copied. The passed memory must be kept alive through message
+  /// processing.
+  /// \param[in] size raw data size.
+  /// \return Status
+  Status Consume(const uint8_t* data, int64_t size);
+
+  /// \brief Feed data to the emitter as a Buffer.
+  ///
+  /// If the emitter can read one or more messages by the Buffer, the
+  /// emitter emits read messages by calling receiver->Receive()
+  /// multiple times.
+  ///
+  /// \param[in] buffer a Buffer to be processed.
+  /// \return Status
+  Status Consume(std::shared_ptr<Buffer> buffer);
```

Comment:
Yes.
If the given buffer doesn't have enough data, emitter keeps the buffer in `chunks_` instead of using it immediately. If we doesn't retain ownership of the given buffer, the buffer may be destructed when emitter uses it.